src/storage/src/hummock/compactor/delete_range_aggregator.rs

```diff
@@ -0,0 +1,242 @@
+                    tombstone.start_user_key = new_tombstone.end_user_key.clone();
+                    tombstone_index.push(new_tombstone);
+                    tombstone_index.push(tombstone);
+                    continue;
+                }
+            }
+            tombstone_index.push(tombstone);
+        }
+
+        DeleteRangeTombstoneIterator {
+            tombstone_index,
+            seek_idx: 0,
+            watermark: self.watermark,
+        }
+    }
+
+    // split ranges to make sure they locate in [smallest_user_key, largest_user_key)
+    pub fn get_tombstone_between(
+        &self,
+        smallest_user_key: &[u8],
+        largest_user_key: &[u8],
+    ) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let mut delete_ranges = vec![];
+        for tombstone in &self.delete_tombstones {
+            if !largest_user_key.is_empty()
+                && tombstone.start_user_key.as_slice().ge(largest_user_key)
+            {
+                continue;
+            }
+
+            if !smallest_user_key.is_empty()
+                && tombstone.end_user_key.as_slice().le(smallest_user_key)
+            {
+                continue;
+            }
+
+            if self.gc_delete_keys && tombstone.sequence <= self.watermark {
+                continue;
+            }
+
+            let start_key = if smallest_user_key.is_empty()
+                || tombstone.start_user_key.as_slice().gt(smallest_user_key)
+            {
+                key_with_epoch(tombstone.start_user_key.clone(), tombstone.sequence)
+            } else {
+                key_with_epoch(smallest_user_key.to_vec(), tombstone.sequence)
+            };
+            let end_key = if largest_user_key.is_empty()
+                || tombstone.end_user_key.as_slice().lt(largest_user_key)
+            {
+                tombstone.end_user_key.clone()
+            } else {
+                largest_user_key.to_vec()
+            };
+            delete_ranges.push((start_key, end_key));
+        }
+        delete_ranges
+    }
+}
+
+pub struct DeleteRangeTombstoneIterator {
+    seek_idx: usize,
+    tombstone_index: Vec<DeleteRangeTombstone>,
+    watermark: u64,
+}
+
+impl DeleteRangeTombstoneIterator {
+    pub fn should_delete(&mut self, user_key: &[u8], epoch: HummockEpoch) -> bool {
+        while self.seek_idx < self.tombstone_index.len()
+            && self.tombstone_index[self.seek_idx]
+                .end_user_key
+                .as_slice()
+                .le(user_key)
+        {
+            self.seek_idx += 1;
+        }
+        if self.seek_idx >= self.tombstone_index.len() {
+            return false;
+        }
+        self.tombstone_index[self.seek_idx]
+            .start_user_key
+            .as_slice()
+            .le(user_key)
+            && self.tombstone_index[self.seek_idx].sequence >= epoch
+            && self.tombstone_index[self.seek_idx].sequence <= self.watermark
```

Comment:
incorrect.
Suppose there are two tombstones X and Y.
key.epoch < X.seq < watermark < Y.seq
However, part of Y is replaced by X in `iter`
Therefore this key will not be deleted by Y, thus this key will not be deleted.
However, this key should have been deleted by X.

Reply:
I have comment this condition. Delay deletion is still correct.

Reply:
see cases in unit-test `test_delete_range_aggregator`. There is a `TODO` comment.